### DISCUSSION TITLE
Suggested Workflow: Templates in Forge Modules

### DISCUSSION THREAD

**Discussion by michaeltlombardi** [puppetlabs-toy-chest/pct#78] [Ideas]
I've put up an [experimental item template for adding a new Resource API type and provider which depends on the PowerShell Manager](https://github.com/puppetlabs/ruby-pwsh/pull/149) from the [`ruby-pwsh`/`puppetlabs-pwshlib` gem/module](https://github.com/puppetlabs/ruby-pwsh).

It will likely soon include an experimental project template for the minimum scaffolding for a Puppet module which has the necessary gems and fixtures.

A workflow that I would like to explore is having the PDK find templates in an expected location of my installed modules (in my experiment, `module_name/pdk/templates`) without having to specify `--templatepath` on each use.

So if that experiment was merged, I would like the following workflow to "just work"

```
puppet module install puppetlabs-pwshlib
pct new rsapi-pwsh-module --name foo
pct new rsapi-pwsh-resource --name foo_bar
```

Right now, I need to specify `--templatepath` to the `pdk/templates` folder inside of wherever `puppetlabs-pwshlib` was installed.

**Comment by michaeltlombardi** [comment]
I don't know how frequently this need will come up, but the `ruby-pwsh`/`puppetlabs-pwshlib` is a library project meant to make it easier to use PowerShell inside of Puppet resources; so it makes sense to me to be able to publish PDK templates which correspond to that specific functionality.

I can similarly envision a library module for python, docker, or even templates for stdlib - any Puppet module which introduces a library of helpers to be consumed by other modules.

**Comment by da-ar** [comment]
At the moment user config is restricted to `~/.pdk/pct.yml` as documented [here](https://github.com/puppetlabs/pdkgo#overriding-template-defaults). I liked your idea around being able to locate that within a "project" of some kind.

Taking that concept, the current thought is that pct would look in the directory it is being run from, say `my_projects/module_name` and override "globally" set user config for the specified templates. If we expand that config file to allow setting a `templatepath` then any time `pct new` is called it would pick up the path from the project configuration. 

I believe that would provide the flexibility without having to think in terms of modules. It could also allow for the shipping of a configuration file with predefined overrides.

**Comment by jpogran** [comment]
The `~/.pdk/pct.yml` config file accepts a `templatepath` right now. When we add the `project` level config, it will still work there.

I am not sure about the 'storing the template in a puppet module' type distribution. We are working to make the templates installable from the forge, so that is where I would expect to find your template. However I could see the case of keeping the template inside your module to keep the points of change together.

**Reply by michaeltlombardi** [reply]
Does the `templatepath` parameter in the `~/.pdk/pct.yml` accept an array of paths? It would be straightforward to add doc notes like

> If you'd like to use the templates included in this module, make sure to add the path to the `pdk/templates` folder in this module to the `templatepaths` section of your `~/.pdk/pct.yml` config file.
> You can quickly grab that path by...

I'm not against maintaining the templates separately, it'll just be another (small) repo to care and feed.
